### PR TITLE
Hubspot Form block base implementation

### PR DESCRIPTION
**Pull request by GP-Dan-Tovbein**
**Description**
---
It's about the new custom [Hubspot Form Block](https://jira.greenpeace.org/browse/PLANET-6190).

For now is only implemented as the default the [Image Full Width](https://p4-designsystem.greenpeace.org/05f6e9516/p/213df0-hubspot-forms/b/99e047) form, but easy to be scalable. 
The rest of the forms will be added once the base implementation is approved. 

-> [Demo page](https://www-dev.greenpeace.org/test-deimos/hubspot-form-block)
-> [Design](https://p4-designsystem.greenpeace.org/05f6e9516/p/213df0-hubspot-forms/b/99e047)

Ref: [Ticket](https://jira.greenpeace.org/browse/PLANET-6190)

If you want to create a new **Hubspot Form** please go [here](https://www-dev.greenpeace.org/dmptest/wp-admin/admin.php?page=leadin_forms&leadin_route%5B0%5D=type&leadin_search=businessUnitId%3D%26dateType%3DLAST_THIRTY_DAYS%26formType%3DALL%26offset%3D0%26parentId%3D0%26query%3D%26sortDirection%3DDESCENDING%26sortingColumn%3DLAST_MODIFIED)

:warning: **Things to keep in mind**
---
- The custom form design will be covered on [this](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/662) pull request.

**Testing**
---
- Make sure you already have a portal id and id related to a Hubspot from before starting.
- Enable `beta` blocks from P4 settings.

**Related PRs** (are in draft)
---
- Use with custom hubspot form styles (overriding the default ones): https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/662
- Implement the Thank you section. https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/663

**Updates**
---
- **September, 9:**
  - Implemented a custom hook to create dynamically the form instead to adding this logic inside the Compoment.
  - Enable to show a custom _Thank you message_ or show the one retrieved by the API.
  - Improve the Hubspot shortcode parser. It can parse 3 different shortcodes (with or without `' and "`)
  ```
  [hubspot type=form portal=8710305 id=85158a32-72b6-4dd7-8012-f4f4d8ef08d1]
  [hubspot type='form' portal='8710305' id='85158a32-72b6-4dd7-8012-f4f4d8ef08d1']
  [hubspot type="form" portal="8710305" id="85158a32-72b6-4dd7-8012-f4f4d8ef08d1"]
  ```
  - ~Show a preloader when the Hubspot form is fetched.~

---

### CODE REVIEW

**Comment by planet-4** [comment]
### Test instance is ready :rocket:

:new_moon: [deimos](test-deimos) | [admin](https://www-dev.greenpeace.org/test-deimos/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-deimos/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-deimos) | [composer-local.json](https://github.com/greenpeace/planet4-test-deimos/blob/main/composer-local.json)

:watch: 2021.08.12 12:03:21

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:37]: Why do we have to use `useMemo` here?

File: assets/src/blocks/HubspotForm/HubspotFormBlock.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormBlock.js:22]: We don't use class components in new code any more. Can you change this to a function component?

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:37]: I see it's used pretty much be default. While `useMemo` **can** in some cases be used to address a performance issue if there's a particularly expensive calculation, in most cases it actually reduces performance. Please check  https://medium.com/swlh/should-you-use-usememo-in-react-a-benchmarked-analysis-159faf6609b7 to get an idea of the complexity needed in the calculation for `useMemo` to actually improve performance.

File: assets/src/blocks/HubspotForm/HubspotFormContext.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormContext.js:40]: Why do we need an effect to run on every render to update the block height?

File: assets/src/editorIndex.js
Comment (Inwerpsel) [inline assets/src/editorIndex.js:29]: Why do we need to wrap this in a try-catch block?

File: assets/src/styles/blocks/HubspotForm/HubspotFormsOverrides.scss
Comment (Inwerpsel) [inline assets/src/styles/blocks/HubspotForm/HubspotFormsOverrides.scss:1]: Can't we include the `hbspt-frm` class without putting it in another class?

File: assets/src/blocks/HubspotForm/HubspotFormThankYou.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormThankYou.js:24]: Please check your IDE settings for blank lines at the end of files :pray:

File: assets/src/blocks/HubspotForm/HubspotFormContext.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormContext.js:40]: Not on every render, this effect is invoked only when the wrapperRef change. In this case only once or twice but no more.

Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormContext.js:40]: Ah sorry I misread the dependencies.

File: assets/src/editorIndex.js
Comment (GP-Dan-Tovbein) [reply assets/src/editorIndex.js:29]: It was suggested by SonarCloud

File: assets/src/blocks/HubspotForm/HubspotFormContext.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormContext.js:7]: Moved to a hook. However, I left the Context only for taking and returning props instead of passing them through each children `Component`.

File: classes/blocks/class-hubspotform.php
Comment (lithrel) [inline classes/blocks/class-hubspotform.php:43]: You can fix your code style issues automatically with phpcs/phpcbf if you have a running instance on your machine:
- Enter the php container `make php-shell`
- Get into the plugin dir `cd public/wp-content/plugins/planet4-plugin-gutenberg-blocks/`
- `./vendor/bin/phpcs` will show you the issues
- `./vendor/bin/phpcbf` will apply fixes

**Review by Inwerpsel — COMMENTED** [review]
Still had these old comments I hadn't submitted, I think they're still relevant.

File: assets/src/blocks/HubspotForm/useDynamicHeight.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/useDynamicHeight.js:3]: Can you explain the motivation for this hook? It's not obvious why this is needed, or if it's needed at all.

The current usage sets the height on an element, where the passed in `wrapperRef` is the only descendant of that element. Doesn't that mean the parent element already has the right height?

File: assets/src/editorIndex.js
Comment (Inwerpsel) [reply assets/src/editorIndex.js]: I think it only tests PRs, so probably it's not historically available for previous commits.

File: assets/src/blocks/HubspotForm/HubspotFormContext.js
Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormContext.js:7]: As discussed on Slack, we should avoid using Context here for now because there's no benefit to using it, it's used in a single place. Context only has benefits for props that get passed deeply, which shouldn't be the case for a block's attributes.

File: assets/src/blocks/HubspotForm/HubspotFormBlock.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormBlock.js:96]: As discussed on Slack, InnerBlocks doesn't allow us the behavior we want, because it only works well if the block that contains it is also saved. In our case the block needs to be rendered, or hydrated. Neither is compatible yet with InnerBlocks.

However we can achieve the same goal (using a block editor for the thank you message) in a much simpler way: By creating a separate block for the thank you message, which is hidden by default on the page, and then shown by the form submit handler

File: assets/src/blocks/HubspotForm/useDynamicHeight.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/useDynamicHeight.js]: You're are right, I've removed it. 
This hook has been used before, now it's not necessary for the current implementation.

File: assets/src/blocks/HubspotForm/useCreateHubspotForm.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/useCreateHubspotForm.js]: This block has some tricky implementations. One is that the user can show the submitted message provided by the plugin or use a custom one. Then, we need to detect when the form is submitted.

I've changed these listeners to this [piece of code](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/641/files#diff-a4a59ac7034e42b21e90e1dafbe268a2677eb4be25699b685b1052ddb6a94800R15-R20).

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:13]: If this is the main background for the block, we could just use `backgroundImage`.

These attributes can be camelCase, btw. I think the snake_case format in our code is a leftover from the shortcodes. WordPress itself nowadays uses camelCase for all block attributes.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:30]: This ref now seems unused, can probably be removed.

**Review by Inwerpsel — COMMENTED** [review]
I had this other comment still pending about `InnerBlocks`, and 2 small new ones.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:30]: It's used [here](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/641/files/02445295ca2c4f40f40f94e456928e42cccd957a#diff-74590d4ebb9c6d57f60a788911755629e1110b04dd2b87c5a18c4c963f26c261R83)

Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:30]: Well the reference is passed there, but nothing actually uses it, right?

Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:30]: Yes, I see your point and you're completely right. It was used by a removed hook. Good catch!

**Comment by GP-Dan-Tovbein** [comment]
@lithrel @Inwerpsel FYI, I've flagged to UAT Passed

File: assets/src/blocks/HubspotForm/HubspotFormSidebar.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormSidebar.js:44]: Does this function name still makes sense? It doesn't seem to ever render an image 🤔

File: assets/src/blocks/HubspotForm/HubspotFormSidebar.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormSidebar.js:8]: It's not a big deal, but there's a small typo here, it should be `Wrapper` with 2 p's 🙂

File: classes/blocks/class-hubspotform.php
Comment (mleray) [inline classes/blocks/class-hubspotform.php:89]: I don't think that we need Hammer for this block? 🤔

File: webpack.config.js
Comment (mleray) [inline webpack.config.js:71]: Don't we also need to add the CSS files here in the config?

File: assets/src/blocks/HubspotForm/HubspotFormBlock.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormBlock.js:86]: I believe this can be shortened to just `edit: HubspotFormEditor,` 🙂

File: assets/src/blocks/HubspotForm/hooks/useBackgroundImage.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/hooks/useBackgroundImage.js:10]: Why do we return an array here? 🤔 It seems it would be simpler to just return the value

File: assets/src/blocks/HubspotForm/HubspotFormFrontend.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormFrontend.js:36]: Is it necessary to use the backgroundImage twice (here and on the `section` element above)?

File: assets/src/blocks/HubspotForm/hooks/useToAttribute.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/hooks/useToAttribute.js:22]: Same comment here as earlier, why do we return an array?

File: webpack.config.js
Comment (GP-Dan-Tovbein) [reply webpack.config.js:71]: What's the main reason to add it also in the config?

Comment (mleray) [reply webpack.config.js:71]: We do it for other blocks in the [cssConfig](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/79bc8812ac1b9cfa4e12ef77c5b288c0d489efb5/webpack.config.js#L74), I thought it was needed to properly split up the blocks' code

Comment (GP-Dan-Tovbein) [reply webpack.config.js:71]: Gotcha! I'll move it.

Just to know, what's the difference between adding to Webpack.config instead of [blocks.scss](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/641/files#diff-cbca468115dd29ef21fbf296256aaf045b3c5fa6d03d09caa7a41ff222f315d4R12)?

Comment (mleray) [reply webpack.config.js:71]: I might be wrong, but I think that when adding them there we only load the styles when the block is actually present on the page, instead of always loading them - so it's better for performance. But don't quote me on that 😬 @Inwerpsel probably knows better!

File: assets/src/blocks/HubspotForm/HubspotFormFrontend.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormFrontend.js:36]: Yes, since its layout is different between large and small screens.
Large screens -> The background image is set to the main wrapper (including block text and block form)
Smaller screens -> The background image is only set to the block text.

Have a look the [design](https://p4-designsystem.greenpeace.org/05f6e9516/p/213df0-hubspot-forms/b/99e047)

File: assets/src/blocks/HubspotForm/HubspotFormBlock.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormBlock.js:86]: Yes, you're right. This is because I thought to potentially render the thank you message editor through this method.

File: webpack.config.js
Comment (Inwerpsel) [reply webpack.config.js:71]: Entirely correct @mleray

File: assets/src/blocks/HubspotForm/hooks/useHubspotForm.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/hooks/useHubspotForm.js]: @Inwerpsel unfortunately your suggestion is not possible since this method is calling another function `createHubspotForm` which is also inside the hook and is on charge of modifying `submitted` and `submittedMessage` hook state variables and this is why I added inside the hook.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: We'll need this functionality in other blocks as well, so maybe we could make it into a separate component already? 🤔 Ofc actually using it in other blocks should be taken care of in another ticket

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:65]: You might not need to add the `container` class to the block anymore, since we recently merged changes removing that class from blocks and adding it to the page content element instead 🤔

File: assets/src/styles/blocks/HubspotForm/HubspotFormEditorStyles.scss
Comment (mleray) [inline assets/src/styles/blocks/HubspotForm/HubspotFormEditorStyles.scss:11]: If you do end up removing the `container` class from the block, you won't need this anymore

File: assets/src/styles/blocks/HubspotForm/HubspotFormStyles.scss
Comment (mleray) [inline assets/src/styles/blocks/HubspotForm/HubspotFormStyles.scss:24]: Same comment here as in the other css file, about removing the `container` class 🙂

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (mleray) [inline assets/src/blocks/HubspotForm/HubspotFormEditor.js:69]: Shouldn't the title also have a max length? 🤔

**Review by mleray — CHANGES_REQUESTED** [review]
I've tested the block on the test instance, and I've noticed a couple of things:
- When the sidebar is open, the block preview goes out of the screen
Screenshot 2021-12-08 at 14 00 52

- I find the default text for the custom "Thank you" message a bit strange, I feel like the placeholder alone would be better but ofc the design team should be consulted on that matter 🙂 
Screenshot 2021-12-08 at 14 01 40
Screenshot 2021-12-08 at 14 01 49

- When there's no background image, it just shows a black background but we actually do have an [ImagePlaceholderIcon component](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/master/assets/src/components/ImagePlaceholderIcon.js) that we could use instead? It would be more consistent with other blocks such as the CarouselHeader. Also, it's a small thing but the "Remove" button should be disabled when there is no background image selected, as it might be confusing otherwise

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: I completely agree with this but I suggest to merge this implementation as a base one and then continue working on improvements in the future.

Comment (mleray) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: I see your point, I'm just worried that we'll forget this component is defined here when we start implementing this for other blocks... Anyway, it's not a blocker, so I'll leave it up to you 🙂

Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: I agree it should probably be a separate component, but let's do that in a follow up so we don't have to rush it.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (Inwerpsel) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:65]: Agree, but if it requires a lot of CSS changes we can also do it in a follow up.

File: assets/src/blocks/HubspotForm/hooks/useToAttribute.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/hooks/useToAttribute.js:11]: We probably don't want to enforce it this way.

What if someone copy pastes text into the block from somewhere? Also if you would start changing a word in the middle of your sentence, it will disappear at the end and won't come back if you delete the extra characters again.

If the text is too long, we should show them a warning, and optionally block the post from saving until they fixed it. We can do this using validation on attributes, [like for slides on the Carousel Header block](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/9dd382fc04caa03d56f62b4038e21f9dac3d3cd1/assets/src/blocks/CarouselHeader/CarouselHeaderBlock.js#L26-L35).

But it's also something we can do in a follow up. Definitely should be fixed before using the block in production.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (mleray) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:65]: Indeed, it's not a blocker!

File: assets/src/blocks/HubspotForm/hooks/useBackgroundImage.js
Comment (Inwerpsel) [inline assets/src/blocks/HubspotForm/hooks/useBackgroundImage.js:7]: You could replace this whole function with just this:
```
export const useBackgroundImage = (image) => {
  return { backgroundImage: image ? `url(${image})` : 'none' };
}
```
and it would work just as well, only the first time it wouldn't return `null` for the background.

String interpolation is a very cheap operation. Much cheaper than the overhead the `useState` and `useEffect` hooks add.

Using an effect would only make sense here if you called an API to get the specific background each time, or do some heavy calculation.

**Review by Inwerpsel — APPROVED** [review]
Left a few more comments and suggestions but nothing that blocks us from merging it as a beta block.

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (comzeradd) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: @GP-Dan-Tovbein can you make a follow up ticket to start capturing all the things we need to work on afterwards in order to put the block out of beta?

**Comment by comzeradd** [comment]
@GP-Dan-Tovbein rebasing will also fix the failing Acceptance tests

File: assets/src/blocks/HubspotForm/HubspotFormEditor.js
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/HubspotForm/HubspotFormEditor.js:14]: For sure @comzeradd

**Comment by GP-Dan-Tovbein** [comment]
I've created this [ticket](https://jira.greenpeace.org/browse/PLANET-6504) to follow up all pending stuffs of coming iterations.

c.c. @mleray feel free to "approve" your requested changes that will allow me to merge this 1st iteration.

**Review by mleray — APPROVED** [review]
Looks good enough to be released as beta 👍 However the commits probably need to be squashed before merge? 🤔

**Pull request merged**
